### PR TITLE
nixos/prometheus: add mqtt-exporter

### PR DESCRIPTION
**Pull request by christoph-heiss**
## Description of changes

~~This adds [mqtt-exporter](https://github.com/kpetremann/mqtt-exporter/), a MQTT exporter for Prometheus, as a package itself.~~
Adds a corresponding `services.prometheus.exporters.mqtt` exporter, such that it can be easily configured & enabled. As well as a test for it, as it should be.  

~~Depends on #288219.~~

<details>
<summary>Obsoleted by 1.4.2</summary>

~~Two patches are unfortunaly (currently) needed, one for providing some basic pyproject properties to allow it to build successfully, as upstream does not have `[project]` defined in their `pyproject.toml`, as well as not have `setup.{cfg,py}`. We could also manually assemble the package I guess, but that seemed like a pretty clean solution to me - or there is some even better solution?~~

~~Second patch addresses the lack of a configuration option to set the port the metrics exporter should listen on, which is an common option pre-defined for all exporters.~~
 
~~I intend to upstream both patches - in fact, I have already sent a PR upstream with both of them, so hopefully they can be dropped soon.~~

~~Both changes have been merged upstream, so they can be dropped with the next release. :rocket:~~
</details>

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by christoph-heiss** [comment]
Ping :^)

File: nixos/modules/services/monitoring/prometheus/exporters/mqtt.nix [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+      PROMETHEUS_PREFIX = cfg.prometheusPrefix;
+      TOPIC_LABEL = cfg.topicLabel;
+      ZIGBEE2MQTT_AVAILABILITY = toConfigBoolean cfg.zigbee2MqttAvailability;
+      ZWAVE_TOPIC_PREFIX = cfg.zwaveTopicPrefix;
+      ESPHOME_TOPIC_PREFIXES = toConfigList cfg.esphomeTopicPrefixes;
+      HUBITAT_TOPIC_PREFIXES = toConfigList cfg.hubitatTopicPrefixes;
+    };
+    serviceConfig = {
+      EnvironmentFile = mkIf (cfg.environmentFile != null) cfg.environmentFile;
+      ExecStart = "${pkgs.mqtt-exporter}/bin/mqtt-exporter";
```
Comment (FliegendeWurst) [inline nixos/modules/services/monitoring/prometheus/exporters/mqtt.nix:137]:
```suggestion
      ExecStart = "${lib.getExe pkgs.mqtt-exporter}";
```

**Review by FliegendeWurst — COMMENTED** [review]
LGTM

**Comment by christoph-heiss** [comment]
> LGTM

Thanks for the review! Fixed your suggestion.

File: pkgs/by-name/mq/mqtt-exporter/package.nix [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  fetchFromGitHub,
+  python3Packages,
+}:
+
+python3Packages.buildPythonApplication rec {
+  pname = "mqtt-exporter";
+  version = "1.4.2";
+  pyproject = true;
+
+  disabled = python3Packages.pythonOlder "3.8";
```
Comment (FliegendeWurst) [inline pkgs/by-name/mq/mqtt-exporter/package.nix:12]: Oldest Python we have is 3.9
```suggestion
```

**Review by FliegendeWurst — APPROVED** [review]

File: pkgs/by-name/mq/mqtt-exporter/package.nix (hunk 2 above)
Comment (christoph-heiss) [reply pkgs/by-name/mq/mqtt-exporter/package.nix]: Thanks, fixed!

File: pkgs/by-name/mq/mqtt-exporter/package.nix [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    owner = "kpetremann";
+    repo = "mqtt-exporter";
+    rev = "v${version}";
+    hash = "sha256-DjvFm6Om2+iU6/UuLsv/vkssjIB3ZR6Ayz0lfW1Izd4=";
+  };
+
+  propagatedBuildInputs = with python3Packages; [
+    paho-mqtt_2
+    prometheus-client
+    setuptools
```
[the remark this answers is not in the document]
Comment (christoph-heiss) [reply pkgs/by-name/mq/mqtt-exporter/package.nix:24]: Yeah, `nativeBuildInputs` is sufficient for mqtt-exporter. Good catch, thanks!

**Comment by christoph-heiss** [comment]
I've now also added a check phase using `pytestCheckHook`, FWIW.

**Review by FliegendeWurst — APPROVED** [review]
Would be good to have a prometheus maintainer review

**Comment by christoph-heiss** [comment]
Package has been added in #360486 separately.

**Review by misuzu — APPROVED** [review]

**Pull request merged**